README
Common utility library for DAITSS


CHANGELOG
0.1.4 - 2011-10-26

 * reduce default max_lines to 2000

 * add support for multiple message arguments to info, err and warn methods.

0.1.3 - 2011-09-16

 * Extend tests for a common error case, a nil value for a section header, as could be produced by a missing environment variable

 * Allow config object to have additional keys (methods) added if they do not already exist

0.1.2 - 2011-08-25

 * Reworked Rakefile to use new rspec

 * Changed logging output format to be more consistent

 * Added spec tests for file-based logs

0.1.1 - 2011-08-05

 * Moved Logger under Datyl namespace

0.1.0 - 2011-08-18 

 * Moved Config and Reporter classes under Datyl namespace.



spec/config_spec.rb
$LOAD_PATH.unshift File.expand_path(File.join(File.dirname(__FILE__), '..', 'lib'))

require 'datyl/config'


# see the config.yml file in this directory

def test_config_filename
  File.join File.dirname(__FILE__), 'config.yml'
end


describe Config do

  it "should raise an error if the config file doesn't exist"  do
    lambda { Datyl::Config.new('/4fl46p8bjkdf')}.should raise_error(/can't find the specified YAML file/i)
  end

  it "should raise an error if it gets a non string file path (e.g. ENV['CONFIG_FILE'] returns nil)"  do
    lambda { Datyl::Config.new(nil)}.should raise_error(/wasn't supplied a valid YAML file/i)
  end

  it "should correctly parse a valid YAML file" do
    lambda { Datyl::Config.new(test_config_filename, :silos)}.should_not raise_error
  end

  it "should raise an error if no sections are specified" do
    lambda { Datyl::Config.new(test_config_filename)}.should raise_error
  end

  it "should raise an error if a non-existant section specified" do
    lambda { Datyl::Config.new(test_config_filename, :foobar)}.should raise_error(/setup could not find a section named/i)
  end

  it "should raise an error if a section variable is nil-valued" do
    lambda { Datyl::Config.new(test_config_filename, nil) }.should raise_error(/found a section name of class NilClass/)
  end

  it "should not raise an error if a section is empty" do
    lambda { Datyl::Config.new(test_config_filename, :daitss)}.should_not raise_error
  end

  it "should properly find all the keys froma particular section, when the section is named by a string" do
    conf = Datyl::Config.new(test_config_filename, 'silos')
    conf.temp_dir_env.should == '/var/daitss/'
    conf.silo_connection_string.should == 'postgres://silo:mysecret@localhost/test_store_db'
  end

  it "should properly find the keys froma particular section, when the section is named by a FQDN" do
    conf = Datyl::Config.new(test_config_filename, 'store-master.fcla.edu')
    conf.required_pools.should == 2
  end


  it "should properly find all the keys from a particular section, when the section is named by a symbol (programmer styling points)" do
    conf = Datyl::Config.new(test_config_filename, :silos)
    conf.temp_dir_env.should == '/var/daitss/'
    conf.silo_connection_string.should == 'postgres://silo:mysecret@localhost/test_store_db'
  end

  it "should allow methods or array notation for included keys" do
    conf = Datyl::Config.new(test_config_filename, :silos)
    conf.temp_dir_env.should == conf['temp_dir_env']
    conf.silo_connection_string.should == conf['silo_connection_string']
  end

  it "should list all included keys from a configuration" do
    conf = Datyl::Config.new(test_config_filename, :database)
    conf.keys.sort.should == ['daitss_connection_string', 'silo_connection_string', 'storemaster_connection_string' ]
  end

  it "should list all values from a configuration" do
    conf = Datyl::Config.new(test_config_filename, :database)
    conf.values.sort.should == [ 'postgres://daitss:topsecret@localhost/daitss_db', 'postgres://silo:topsecret@localhost/store_db', 'postgres://storemaster:topsecret@localhost/daitss_db' ]
  end

  it "should only find the keys from the specified section, and no more" do
    conf = Datyl::Config.new(test_config_filename, :silos)
    conf.keys.length == 2
    conf.keys.include?('temp_dir_env').should == true
    conf.keys.include?('silo_connection_string').should == true
  end

  it "should get the value of the last instance of multiply defined variables when a section repeats them" do
    conf = Datyl::Config.new(test_config_filename, :silos, :database)
    conf.silo_connection_string.should == 'postgres://silo:topsecret@localhost/store_db'
    conf = Datyl::Config.new(test_config_filename, :database, :silos)
    conf.silo_connection_string.should == 'postgres://silo:mysecret@localhost/test_store_db'
  end
  
  it "should work for funky section and variable names" do   
    lambda { Datyl::Config.new(test_config_filename, 'funky-section') }.should_not raise_error
    
    conf = Datyl::Config.new(test_config_filename, 'funky-section')
    conf['funky-key'].should == 'funky value'
  end

  it "should allow us to iterate over configuration variables" do    
    conf = Datyl::Config.new(test_config_filename, :fixity)
    what_is        = {}
    what_should_be = {  'expiration_days' =>  45,  'stale_days' =>  60 }
    conf.each { |k, v| what_is[k] = v }
    what_is.should == what_should_be
  end

  it "should raise error if an object method is called with arguments" do
    conf = Datyl::Config.new(test_config_filename, :fixity)
    lambda { conf.expiration_days(10)  }.should raise_error(/method expiration_days takes no arguments, but got 1/)
  end

  it "should let us add a new key to an initialized config object" do
    conf = Datyl::Config.new(test_config_filename, :fixity)
    conf['foo'] = :bar

    conf.foo.should == :bar
  end

  it "should not let us add a key to an initialized config object if it already exists" do
    conf = Datyl::Config.new(test_config_filename, :fixity)
    lambda { conf['expiration_days'] = 10  }.should raise_error(/method expiration_days already exists/)
  end

end


spec/logger_spec.rb
$LOAD_PATH.unshift File.expand_path(File.join(File.dirname(__FILE__), '..', 'lib'))

require 'fileutils'
require 'datyl/logger'

def logfile
  File.join(File.dirname(__FILE__), 'test.log')
end

# see http://www.python.org/dev/peps/pep-3333/ for details; this is the basis of the rack spec

def pep333_env
  {
    'PATH_INFO'        => '/index.html',
    'REMOTE_ADDR'      => 'www.example.org',
    'REMOTE_USER'      => 'fischer',
    'REQUEST_METHOD'   => 'GET',
    'SERVER_PROTOCOL'  => 'HTTP/1.1',
    'QUERY_STRING'     => ''
  }
end


# get the last line out of the logfile

def getlog
  # logfilename = `ls  #{File.join(File.dirname(__FILE__), 'test00*.log')}`.split.sort.pop

  File.read(logfile).split("\n").pop.strip
end


describe Datyl::Logger do

  before do
    FileUtils.rm_f logfile()
    Datyl::Logger.setup('RSpec', 'localhost')
    Datyl::Logger.filename = logfile()
  end
 
  it "should allow us to create a log object which writes to file, including severity (INFO) and initialization tag" do
    logger = Datyl::Logger.new(:info, 'My Tag:')
    logger.write('test')
    getlog.should =~ /INFO My Tag: test$/
  end

  it "should include the date in the beginning of the logging output" do
    Datyl::Logger.info 'some stuff'
    getlog.should =~ /^\d\d\d\d\-\d\d\-\d\d \d\d:\d\d:\d\d/
  end

  it "should allow us to write an informational message using the info class method" do
    str = "This is some information"
    Datyl::Logger.info str
    getlog.should =~ /INFO #{str}$/
  end

  it "should allow us to write a warning message using the warn class method" do
    str = "This is a warning"
    Datyl::Logger.warn str
    getlog.should =~ /WARN #{str}$/
  end

  it "should allow us to write an error message using the err class method" do
    str = "This is an error"
    Datyl::Logger.err str
    getlog.should =~ /ERROR #{str}$/
  end

  it "should include apache styling when the proper environment is included" do
    Datyl::Logger.info "foo", pep333_env
    getlog.should =~ %r{www.example.org - fischer "GET /index.html HTTP/1.1"$}
  end
end


spec/reporter_spec.rb
require 'datyl/reporter'
require 'tempfile'

# Test playback functions of reporter - not we are not checking the Logging output part of the Reporter.


$LOAD_PATH.unshift File.join(File.dirname(__FILE__), '..', 'lib')

def output reporter
  output_file = Tempfile.new('rspec-reporter-output')
  output_file.unlink

  reporter.write output_file
  output_file.rewind
  output_file.read
end


describe Datyl::Reporter do
    
  it "should create a report object with a simple title" do

    report = Datyl::Reporter.new "Main Title"
    text = output(report)
    text.should =~ /^Main Title$/n
    text.should =~ /^::::::::::$/n
  end

  it "should create a report object with a title and subtitle" do
    report = Datyl::Reporter.new "Main Title", "Subtitle"
    text = output(report)
    text.should =~ /^Main Title: Subtitle$/n
    text.should =~ /^::::::::::::::::::::$/n
  end

  it "should create a report object with lines of text and a trailing line" do
    report = Datyl::Reporter.new "Main Title", "Subtitle"
    report.err 'An Error'
    report.warn 'A Warning'
    report.info 'An Info'

    text  = output(report)
    lines = text.split("\n")

    lines[2].should == 'An Error'
    lines[3].should == 'A Warning'
    lines[4].should == 'An Info'

    text[-1].chr.should == "\n"   # indicates blank line
    text[-2].chr.should == "\n"
  end

  it "should accept mutliple messages" do
    report = Datyl::Reporter.new "Main Title", "Subtitle"
    report.info 'This', 'is', 'a test'

    text  = output(report)
    lines = text.split("\n")

    lines[2].should == 'This'
    lines[3].should == 'is'
    lines[4].should == 'a test'

    text[-1].chr.should == "\n"   # indicates blank line
    text[-2].chr.should == "\n"
  end

  it "should not be interesting if it is empty" do
    report = Datyl::Reporter.new "Main Title", "Subtitle"
    report.interesting?.should == false
  end    

  it "should be interesting if it has content" do
    report = Datyl::Reporter.new 'Main Title', 'Subtitle'
    report.info 'booga booga'
    report.interesting?.should == true
  end    

  it "should insert blank lines when we 'info' without an argument" do
    Datyl::Reporter.max_lines_to_write = 10
    report = Datyl::Reporter.new 'Main Title', 'Subtitle'

    report.info 'this'
    report.info 
    report.info 'a test'
    
    lines = output(report).split("\n")
    lines[2].should == 'this'
    lines[3].should == ''
    lines[4].should == 'a test'
  end

  it "should clip out extra lines" do
    Datyl::Reporter.max_lines_to_write = 10
    report = Datyl::Reporter.new 'Main Title', 'Subtitle'

    100.times do |num|
      report.info "Line #{num}"
    end

    text = output(report)

    # We're expecting something along the lines of this:

    # Main Title: Subtitle
    # ::::::::::::::::::::
    # Note: 90 of 100 lines were discarded - see the system log for the complete report.
    # Line 0
    # Line 1
    # Line 2
    # Line 3
    # Line 4
    #  ...
    # Line 95
    # Line 96
    # Line 97
    # Line 98
    # Line 99

    lines = text.split("\n")

    lines.shift.should == 'Main Title: Subtitle'
    lines.shift.should == ':' * 20
    
    lines.shift.should =~ /^Note: 90 of 100 lines were discarded/

    lines[0].should  == 'Line 0'
    lines[4].should  == 'Line 4'
    lines[5].should  == ' ...'
    lines[6].should  == 'Line 95'
    lines[10].should == 'Line 99'

    lines.count.should == 11
  end



end


lib/datyl/trie.rb
# -*- coding: utf-8 -*-
#
# Randy Fischer (rf@ufl.edu) reimplemented this wheel on 2010-06-13.  Pleasant exercise.
#
# Build a trie - my goal was to remove the largest common prefix from
# a list of strings. See the twigs method to get that.  This also does
# the usual trie tricks: stores key/value pairs, orders the keys, etc.
#
# Example program:
# 
#  trie = Trie.new
#  
#  trie['silos.darchive.fcla.edu:/daitssfs/001'] = :a
#  trie['silos.darchive.fcla.edu:/daitssfs/002'] = :b
#  trie['silos.darchive.fcla.edu:/daitssfs/002'] = :c
#  trie['silos.darchive.fcla.edu:/daitssfs/004'] = :d
#  trie['silos.darchive.fcla.edu:/daitssfs/010'] = :e
#  trie['silos.darchive.fcla.edu:/daitssfs/015'] = :f
#  trie['silos.darchive.fcla.edu:/daitssfs/027'] = :g
#
#  trie['silos.darchive.fcla.edu:/daitssfs/002']     => :c
#
#  
#  trie.keys   =>  [ "silos.darchive.fcla.edu:/daitssfs/001", "silos.darchive.fcla.edu:/daitssfs/002", .. ]
#  trie.values =>  [ :a, :c, :d, :e, :f, :g ]
#
#  trie.prefix => 'silos.darchive.fcla.edu:/daitssfs/0'
#  
#  trie.twigs  => [ '01', '02', '04', '10', '15', '27' ]

class Trie

  # Box let's us put a value into a box, then get it out of the box.
  # We need this because we'd like to store nil and false values on
  # a node slot.  So a slot can have one of two values: nil (meaning
  # no value), or  a box (with some value in it, including nil).

  class Box

    attr_reader :value

    def initialize thing
      @value = thing
    end

    def unwrap
      value
    end
    
    def to_s
      "#<box##{self.object_id}: #{value.inspect}>"
    end
  end

  # A node stores a single character, pointers to children nodes, and perhaps a boxed value.
  # 
  # Example Node: < letter => 'a', value => 'adjective, also the indefinite article, refers...', children => [ <node>, <node> .. ] >

  class Node

    attr_reader   :letter, :children
    attr_accessor :value

    def initialize letter = ''
      @letter     = letter
      @children   = []
      @value      = nil
    end

    # Create a new child node that stored the string LETTER.  Returns the new node.

    def add_child letter
      new_node = Node.new(letter)
      @children.push new_node
      @children.sort! { |node_a, node_b| node_a.letter <=> node_b.letter }
      new_node
    end
  end

  attr_reader :root

  # Our trie starts with a node with an empty string.

  def initialize
    @root = Node.new
  end

  def [] key
    val = lookup(key)
    val.nil? ? nil : val.unwrap
  end


  def []= key, val
    store(key, Box.new(val))
    return val
  end

  # Return longest common prefix from the trie.

  def prefix node = root, str = ''
    return str + node.letter if node.children.count != 1 or not node.value.nil?
    return prefix node.children[0], str + node.letter
  end

  def keys
    collection = []
    find_keys collection
    collection
  end

  def values
    collection = []
    find_values collection
    collection.map { |box| box.unwrap }
  end
    
  # There's a broken by design bug in here: if the largest common
  # prefix is itself a key, we'll get an empty string. You'll need to
  # check for that case. TODO: fix that.

  def twigs
    len = prefix.length
    keys.map { |k| k[len..-1]}
  end

  # Huh.  Must be for debugging or sumthin...

  def dump node = root, indent = ''
    STDERR.puts indent + node.letter + (node.value.nil? ? "" : " " + node.value.inspect) unless node == root
    node.children.each do |n|
      dump n, indent + '→  '
    end
  end

  private

  # Store a value for a key. Used for []= above.

  def store key, val, node = root

    head = key[0..0]              # divide and conquer
    tail = key[1..-1]

    found = node.children.select { |n| n.letter == head }.first   # look for our letter in the children..
    found ||=  node.add_child head                                # create a node for it if need be.

    if tail.empty?                # we're done; save value
      found.value = val
    else                          # otherwise, keep dividing, keep conquering.
      store tail, val, found
    end
  end

  # Find all keys. Helper for method keys.

  def find_keys collection, node = root, str = ''
    collection.push str + node.letter unless node.value.nil?
    node.children.each do |nd|
      find_keys collection, nd, str + node.letter
    end      
  end

  # Find all values, same order as corresponding keys are
  # returned. Helper for method values.

  def find_values collection, node = root, str = ''
    collection.push node.value unless node.value.nil?
    node.children.each do |nd|
      find_values collection, nd, str + node.letter
    end      
  end

  # Lookup value for a key. Used for []

  def lookup key, node = root
    return nil if key.empty?

    head = key[0..0]
    tail = key[1..-1]

    node.children.each do |nd|
      next unless head == nd.letter
      return nd.value if tail.empty? and not nd.value.nil?
      return lookup tail, nd
    end
    
    return nil
  end

end



lib/datyl/config.rb
require 'yaml'

module Datyl

  # The Config class acts a lot like a simple Struct, with named slots
  # read out of a configration file of key/value pairs. The
  # configuration file is a simple YAML file with specific named
  # sections.  For instance:
  #
  #    database:
  #      daitss_connection_string: postgres://daitss:topsecret@localhost/daitss_db
  #      storemaster_connection_string: postgres://storemaster:topsecret@localhost/daitss_db
  #      silo_connection_string: postgres://silo:topsecret@localhost/store_db
  #
  #    silos.example.org:
  #      temp_dir_env: /var/daitss/
  #      silo_connection_string: postgres://silo:mysecret@localhost/test_store_db
  #
  #       ....
  #
  # Each section contains simple key/value pairs. When we create a new
  # Config object, we specify what sections we're interested in. Each
  # section is processed in the order presented to the constructor, so
  # a key that's found in multiple sections (like the key
  # silo_connection_string above) gets the value of the last key
  # encountered if both the 'database' and 'silos.example.org'
  # sections are included.
  #
  # Typically, we expect a few global sections at the top of the file
  # and specified first in the constructor.


  class Config

    def initialize yaml_path, *sections

      # We try to be very specific with our error messages here, since
      # configuration is a pain point for new installers.

      raise "Configuration setup wasn't supplied a valid YAML file path - instead it got a #{yaml_path.class}" unless yaml_path.class == String
      raise "Configuration setup can't find the specified YAML file #{yaml_path}" unless File.exists? yaml_path
      raise "Configuration setup can't read the specified YAML file #{yaml_path}" unless File.readable? yaml_path
      raise "Configuration setup must be supplied with one or more sections for the yaml file #{yaml_path}" if sections.empty?

      # handles nil value, an easy-to-make error when a missing environment variable is used for a section name:

      sections.each do |sec|
        raise "Configuration setup found a section name of class #{sec.class}, but only strings or symbols are allowed (sections arguments: #{sections.inspect})" unless [String, Symbol].include?(sec.class)
      end

      begin
        @yaml = YAML.load_file yaml_path
      rescue => e
        raise "Configuration setup did not correctly parse the specified YAML file #{yaml_path}: #{e.message}"
      end

      if @yaml.class != Hash
        raise "Configuration setup parsed the specified YAML file #{yaml_path}, but it's not the expected simple hash (it's a #{@yaml.class})"
      end

      @hash = Hash.new

      sections.each do |sect|
        sect = sect.to_s

        data = @yaml[sect]

        next if data.nil? and @yaml.keys.include? sect     # catch empty section

        if data.nil?
          raise "Configuration setup could not find a section named #{sect} in the specified YAML file #{yaml_path}"
        end

        data = @yaml[sect]
        if data.class != Hash
          raise "Configuration setup expected that the section named #{sect} from the specified YAML file #{yaml_path} would be a hash, but instead it's a #{data.class}"
        end
        data.each { |k,v|  @hash[k] = v }
      end
    end

    def method_missing method, *args
      if not args.empty?
        raise "method #{method} takes no arguments, but got #{args.length}"
      end
      return @hash[method.to_s]
    end


    def [] key
      return @hash[key.to_s]
    end

    def keys
      @hash.keys
    end

    def values
      @hash.values
    end

    def each
      @hash.each do |k, v|
        yield k, v
      end
    end

    # it can be convenient to add an accessor to a config object (e.g. to set defaults for
    # missing keys):

    def []= key, value
      raise "method #{key} already exists on this configuration object" if @hash.key? key
      @hash[key.to_s] = value
    end


  end # of class Config
end # of module Datyl


lib/datyl/reporter.rb
require 'datyl/logger'
require 'tempfile'
require 'time'


module Datyl

  # This class combines logging and an abbreviated, written report. The Logger system is assumed to have
  # been initialized (logging will be a no-op otherwise).  
  #
  # Create a reporter object by initializing with a Title and an optional Subtitle.
  # Example:
  #
  #   rep = Reporter.new 'Important Report', 'This could be very important'
  #
  #   rep.info 'This is a test.'      => Logs an info-level message:     'Important Report: This is a test.'
  #   rep.warn 'This is a warning.'   => Logs an warning-level message:  'Important Report: This is a warning.'
  #   rep.err  'This is an error.'    => Logs an error-level message:    'Important Report: This is an error.'
  #
  #   rep.write STDERR   => Produces on STDERR:
  #
  #                         Important Report: This could be very important
  #                         ==============================================
  #                         This is a test.
  #                         This is a warning.
  #                         This is an error.
  #
  # Note: the subtitle is only included in the written report;
  # info/warning/errors are distinguished only in the logs.  The Title
  # is included for each logged line: keep it short - use the Subtitle
  # to expand the heading for the written reports.  Set the class
  # attribute Reporter.max_lines_to_write to limit the number of lines
  # written in any particular report; it will print the first and last
  # lines of the report if the number of lines goes over that value
  # (intially, 1000 lines).  A blank line is added after the text of the
  # written report.
  #

  class Reporter

    @@max_lines = 2000    # will only write this many lines of text - first half, '...', second half.

    attr_reader   :title, :counter

    def initialize title, sub = nil
      @start     = Time.now
      @done      = nil
      @counter   = 0
      @title     = title
      @subtitle  = sub
      @tempfile  = Tempfile.new("report-#{title.split(/\s+/).map{ |word| word.gsub(/[^a-zA-Z0-9]/, '').downcase }.join('-')}-")
      @tempfile.unlink
    rescue => e
      Logger.error "Fatal error in Reporter module; #{e.class}: #{e.message}"
    end

    def self.max_lines_to_write
      @@max_lines
    end

    def self.max_lines_to_write= num
      @@max_lines = num
    end

    # for debugging - if you call 'done' the report will add to the
    # title its total runtime - from the creation time of the
    # constructor to the point 'done' was called

    def done
      @done = Time.now
    end

    # Support for logging messages at various severity levels; slightly abbreviated data is recorded to a playback
    # file.  Called with no arguments, just insert a blank line the playback file.

    def dolog *strs
      @counter += 1
      
      strs.each do |str|
        yield @title + ': ' + str unless str.empty?
        @tempfile.puts str
      end

      @tempfile.puts('') if strs.empty?
    end

    # Log messages using the INFO severity

    def info *strs
      dolog(*strs) { |s| Logger.info s }
    end

    # Log messages using the WARN severity

    def warn *strs
      dolog(*strs) { |s| Logger.warn s }
    end

    # Log messages using the ERR severity

    def err *strs
      dolog(*strs) { |s| Logger.err s }
    end


    def rewind
      @tempfile.rewind
    end

    # interesting? is true if anything has been reported - some reports could be empty

    def interesting?
      @counter > 0
    end

    def _top_lines
      @@max_lines / 2 + (@@max_lines.odd? ? 1 : 0)
    end

    def _bottom_lines
      @@max_lines / 2
    end

    # Cycle through everything that's been written, yielding a line;
    # if the report is large, the *middle* of the report is truncated.
    # It's meant to be used in email reports. Only a total @@max_lines
    # is provided split equally from from the the begginning and the
    # end.

    def each
      title = @title + (@subtitle ? ": #{@subtitle}" : '') +  (@done ? sprintf(" (%3.2f seconds)", @done - @start) : '')
      yield title
      yield title.gsub(/./, ':')

      @tempfile.rewind

      if @counter > @@max_lines
        yield "Note: #{@counter - @@max_lines} of #{@counter} lines were discarded - see the system log for the complete report."

        _top_lines.times  { yield @tempfile.gets }           # print first half

        @tempfile.rewind

        (@counter - _bottom_lines).times { @tempfile.gets }  # discard middle

        yield " ..."

        while not @tempfile.eof                             # print last half
          yield @tempfile.gets
        end

      else                     # we can print everything
        @tempfile.rewind
        while not @tempfile.eof
          yield @tempfile.gets
        end
      end

      yield ''
    end


    # Write the report to the provided IO object, STDOUT by default.
    # Only @@max_lines lines of text are written, split equally
    # between the beggining and end of the report, with an elipsis
    # inserted in between.

    def write io = STDOUT
      each do |line|
        io.puts line
      end
    end

    # A class version of the above which is immediately written out -
    # it's a convenience routine.

    def self.note message, io = STDOUT
      Logger.info message
      io.puts message
    end


  end # of class Reporter
end # of module Datyl 


lib/datyl/logger.rb
require 'log4r'
require 'log4r/outputter/rollingfileoutputter'
require 'log4r/outputter/fileoutputter'
require 'log4r/outputter/syslogoutputter'
require 'socket'
require 'syslog'

module Datyl

  # Initial Author: Randy Fischer (rf@ufl.edu) for DAITSS
  #
  # Logging web service actions using using log4r, along with
  # Rack::CommonLogger and DataMapper::Logger tie-ins.
  #
  # Example use:
  #
  #  require 'logger'
  #
  #  Logger.setup("XmlResolution", "xmlresolution.example.com")
  #  Logger.filename = "myfile.log"
  #  ...
  #  get '/tmp' do
  #     Logger.warn "can't create temporary file for request.", @env
  #     halt [ 500, {}, 'whoops' ]
  #  end
  #  ...
  #  get '/temp' do
  #     Logger.err  "the machine room appears to be on fire.", @env
  #     "451 degrees Fahrenheit"
  #  end
  #
  # This might produce in "myfile.log" the following:
  #
  #   WARN XmlResolution: "GET /tmp" can't create temporary file for request.
  #  ERROR XmlResolution: "GET /temp" the machine room appears to be on fire.
  #
  # The env variable is a hash, typically you'll use @env from the
  # Sinatra environment.  Anything that provides the PEP 333 standard
  # set of CGI environment variables is best (HTTP_SERVER, PATH_INFO,
  # etc), but an empty hash will work in a pinch (and is the default).
  #
  # While we usually use the class methods, there exists support to
  # create a logging object, which can be useful for those libraries
  # that can support redirecting diagnostic messages.
  #
  # Here's an example of using a Logger object to direct
  # Rack::CommonLogger and DataMapper::Logger to write to both a file
  # and syslog.
  #
  #  require 'logger'
  #
  #  Logger.setup('XmlResolutionService', 'xmlresolution.example.com')
  #  Logger.filename = '/tmp/myfile.log'
  #  Logger.facility = :LOG_LOCAL2
  #
  #  use Rack::CommonLogger, Logger.new
  #  DataMapper::Logger.new(Logger.new, :debug)
  #
  # Note a gotcha in DataMapper logging: you must call it before any
  # DataMapper initializations take effect.


  class Logger

    # This class variable lets us know if is safe to proceed; a Log4r
    # logging object has been set up and registered.  It must be set by
    # a call to Logger.setup, followed by one or more of the log output
    # initializers, namely Logger.filename=, Logger.facility= or
    # Logger.stderr.

    @@virtual_hostname = nil
    @@service_name     = nil

    # Logger.setup service_name, virtual_hostname
    #
    # Intialize the logging system.  Until called, all logging is
    # a no-op.
    #
    # In logging the arguments are used as follows:
    #
    # Dec 15 12:38:26 romeo-foxtrot SiloPool[51470]:  INFO silos.test.fcla.edu
    #                               --------               -------------------
    #                               @@service_name         @@virtual_hostname
    #
    # It is recommented the virtual host name of the service be used for
    # @@virtual_hostname but it will default to the hostname of the
    # computer; @@service_name should not have any spaces, so the logs
    # can be conveniently parsed.


    def Logger.setup(service_name, virtual_hostname = nil)
      @@virtual_hostname = virtual_hostname || Socket.gethostname
      @@service_name     = service_name

      Log4r::Logger.new @@virtual_hostname
    end

    # Logger.filename = FILEPATH
    #
    # Intialize the logging system to write to the file named by
    # string FILEPATH.
    #
    # We envision a number of ways to initialize this class before
    # use. We can set up for logging to files, syslog, stderr, or any
    # combination. At least one of them must be done before any actual
    # logging occurs.

    def Logger.filename= filepath
      return unless (@@virtual_hostname and @@service_name)
      formatter = Log4r::PatternFormatter.new(:pattern => "%d %l %m")
      Log4r::Logger[@@virtual_hostname].add Log4r::FileOutputter.new(@@service_name, { :filename => filepath, :trunc => false, :formatter => formatter })
      # Log4r::Logger[@@virtual_hostname].add Log4r::RollingFileOutputter.new(@@service_name, { :maxsize => 100 ,  :filename => filepath, :trunc => false, :formatter => formatter })
      filepath
    end

    # Logger.stderr
    #
    # Initialize the logging system to write to STDERR.

    def Logger.stderr
      return unless (@@virtual_hostname and @@service_name)
      formatter = Log4r::PatternFormatter.new(:pattern => "%d %l %m")
      Log4r::Logger[@@virtual_hostname].add Log4r::StderrOutputter.new(@@service_name, :formatter => formatter)
    end

    # Logger.facility = FACILITY
    #
    # Intialize the logging system to write to syslog, using the
    # the symbol FACILTIY to specify the facility to use. Normally
    # one uses a local facility code (but strings will work):
    #
    #    :LOG_LOCAL0
    #    :LOG_LOCAL1
    #       ...
    #    :LOG_LOCAL9
    #
    # A typical syslog.conf entry using :LOG_LOCAL2 might look like this:
    #
    #    local2.error   /var/log/xmlresolution.error.log
    #    local2.warn    /var/log/xmlresolution.warn.log
    #    local2.info    /var/log/xmlresolution.info.log
    #
    # In the case of Mac OSX Snow Leopard, the above syslog.conf entry
    # means that a Logger.error(message) will get logged to
    # xmlresolution.error.log, xmlresolution.warn.log, and
    # xmlresolution.info.log.  Logger.warn(message) gets logged to
    # xmlresolution.warn.log and xmlresolution.info.log.
    # Logger.info(message) will go to xmlresolution.info.log.  Other
    # syslog deamons on other systems and alternative configurations
    # behave significantly differently.  But that's not your problem.

    def Logger.facility= facility
      return unless (@@virtual_hostname and @@service_name)
      facility = 'LOG_' + facility unless facility =~ /^LOG_/
      Log4r::Logger[@@virtual_hostname].add Log4r::SyslogOutputter.new(@@service_name, 'facility' => eval("Syslog::#{facility.to_s.upcase}"))
      facility
    end

    def Logger.format_message message, env
      suffix = apache_styling(env).strip 
      prefix = message.strip

      return (prefix.empty? or suffix.empty?) ? prefix + suffix : prefix + ' ' + suffix
    end


    # Logger.err MESSAGE, [ ENV ]
    #
    # Log an error message MESSAGE, a string; The hash
    # ENV is typically the Sinatra @env object, but any hash with the
    # common PEP 333 keys could be used.

    def Logger.err message, env = {}
      return unless (@@virtual_hostname and @@service_name)
      Log4r::Logger[@@virtual_hostname].error format_message(message, env)
    end

    # Logger.warn MESSAGE, [ ENV ]
    #
    # Log a warning message MESSAGE, a string; The hash
    # ENV is typically the Sinatra @env object, but any hash with the
    # common PEP 333 keys could be used.

    def Logger.warn message, env = {}
      return unless (@@virtual_hostname and @@service_name)
      Log4r::Logger[@@virtual_hostname].warn format_message(message, env)
    end

    # Logger.info MESSAGE, [ ENV ]
    #
    # Log an informative message MESSAGE, a string; The hash
    # ENV is typically the Sinatra @env object, but any hash with the
    # common PEP 333 keys could be used.

    def Logger.info message, env = {}
      return unless (@@virtual_hostname and @@service_name)
      Log4r::Logger[@@virtual_hostname].info format_message(message, env)
    end

    # While we normally use the class methods to write our own log
    # entries, we also have an object we can instantiate for
    # Rack::CommonLogger or DataMapper::Logger to use. For example:
    #
    #  Logger.setup('XmlResolutionService', 'xrez.example.com')
    #  Logger.filename = 'my.log'
    #  use Rack:CommonLogger, Logger.new(:info, 'Rack:')
    #  DataMapper::Logger(Logger.new(:info, 'DataMapper:'), :debug)

    @level = nil
    @tag   = nil

    def initialize level = :info, tag = ''
      raise "Bad argument: if specified, the first argument must be one of :info, :warn, :error, but was :#{level}" unless [:info, :warn, :error].include? level
      raise "Bad argument: if specified, the second argument must be a simple string, but was a #{tag.class}" unless tag.class == String
      raise "The logging system has not been setup: use Logger.setup(service, hostname)." if (@@virtual_hostname.nil? or @@service_name.nil?)
      @level = level
      @tag   = tag.length > 0 ? tag : nil
    end

    # log.write MESSAGE
    #
    # Rack::CommonLogger can be told to use as a logger any object
    # that has a write method on it. See the Logger#new method for an
    # example of its use.

    def write message
      Log4r::Logger[@@virtual_hostname].send(@level, @tag ? "#{@tag} #{message.chomp}" : message.chomp)
    end

    private

    # apache_styling ENV
    #
    # Create an apache-style "METHOD /uri" string, if possible, from
    # the environment ENV.  ENV is expected to be @env from a rack
    # application or similar.  Stolen from rack's commonlogger with
    # minor corrections.
    #
    # Note that the logger object's #write method will not use this,
    # precisely because we usually use the logger object with rack,
    # which appends exactly this information.

    def Logger.apache_styling env
      return '' if env['REQUEST_METHOD'].nil? or env['PATH_INFO'].nil?

      sprintf('%s - %s "%s %s%s %s"',
              env['HTTP_X_FORWARDED_FOR'] || env["REMOTE_ADDR"] || "-",
              env["REMOTE_USER"] || "-",
              env["REQUEST_METHOD"],
              env["PATH_INFO"],
              env["QUERY_STRING"].empty? ? "" : "?" + env["QUERY_STRING"],
              env["SERVER_PROTOCOL"] || env["HTTP_VERSION"]
              )
    end

  end # of class
end # of module Datyl
